src/django_pg_migration_tools/management/commands/migrate_with_timeouts.py

```diff
@@ -1,15 +1,38 @@
+import dataclasses
 import datetime
-from typing import Any
+import importlib
+import time
+from typing import Any, Callable
 
 from django.core.management import base
 from django.core.management.commands.migrate import Command as DjangoMigrationMC
 
 from django_pg_migration_tools import timeouts
 
 
+@dataclasses.dataclass
+class RetryState:
+    current_exception: timeouts.DBTimeoutError
+    lock_timeouts_count: int
+    statement_timeouts_count: int
+
+
 class Command(DjangoMigrationMC):
+    retry_callback: Callable[[RetryState], None] | None
+
     statement_timeout: datetime.timedelta | None
+    statement_timeouts: int
+    statement_timeouts_max_retries: int
+    statement_timeout_retry_exp: int
+    statement_timeout_retry_max_wait: datetime.timedelta
+    statement_timeout_retry_min_wait: datetime.timedelta
```

Comment:
🤔  There's a strong use case for having retries for lock timeouts, but I'm struggling to see why retrying in case of statement timeouts would be useful. Do we expect a statement that times out the first time to somehow become faster and not to timeout when retried? What's an example of such a statement where that's likely?

Reply:
@timb07 

> Do we expect a statement that times out the first time to somehow become faster and not to timeout when retried?

Good question. I think that this is a possible situation because of temporary load spikes on the database, caused by either background tasks or maintenance tasks starving server resources.

E.g., a vacuum finishing between retries could make the schema change query execute faster and thus pass on a second attempt. Adding a unique constraint to a table with many dead rows is an example of such query. The constraint generation would need to step through many dead-rows and thus have to load more pages from disk to finish its job before the vacuum which could substantially change the results.

I do agree though, that it doesn't feel as useful as lock_timeout retries and its usage might be limited to fewer situations. 

It almost makes sense to add a exponential back-off for setting `statement_timeout` to ever increasing values up to a limit, but that sounds like a different feature.

Reply:
I agree that the examples you've given do fit the scenario I asked about. But they're not terribly convincing as a justification for the need for this feature. Whereas lock timeouts + automatic retries is well-established best practice (e.g. [this blogpost](https://postgres.ai/blog/20210923-zero-downtime-postgres-schema-migrations-lock-timeout-and-retries#graceful-schema-changes-lock_timeout-and-retries)) for applying migrations. 

And let's go back to why we might want a statement timeout for migrations in the first place. The main reason (as I understand it) is that we don't want migrations to take too long, because they delay the overall deployment process. So if we hit the statement timeout, isn't retrying the wrong thing to do?

Reply:
No problems, I will remove this part of the feature and if we find better use cases in the future we can add it back.

Reply:
Addressed in:

- bd5a5ed87fc11223dcef9f096e6e6e31cb5a48d8
- 5c35e23041229870259680286df1b2512d051894